crypto/ocsp/ocsp_test.cc

```diff
@@ -587,39 +770,102 @@ TEST(OCSPTest, TestDefaultHash) {
   ASSERT_TRUE(cert_id);
 
   ASN1_GENERALIZEDTIME *revtime, *thisupd, *nextupd;
-  // Actual verification of the response
+  // Checks revocation status of the response
   const int ocsp_resp_find_status_res = OCSP_resp_find_status(basic_response.get(), cert_id.get(), &status, &reason, &revtime, &thisupd, &nextupd);
   ASSERT_EQ(1, ocsp_resp_find_status_res);
   ASSERT_EQ(V_OCSP_CERTSTATUS_REVOKED, status);
 }
+
+// Test valid OCSP response, but the data itself is untrusted
+TEST(OCSPTest, TestUntrustedDataOCSP) {
+  // Convert const good ocsp response test file to changeable pointer
+  bssl::Span<const uint8_t> der = bssl::Span<const uint8_t>(ocsp_response_der);
+  std::unique_ptr<uint8_t[]> ptr(new uint8_t[der.size()]);
+  OPENSSL_memcpy(ptr.get(), der.data(), der.size());
+
+  // Mess up a byte right in the middle of the cert
+  ptr[800] = ptr[800] + 1;
+  bssl::Span<const uint8_t> malformed_der = bssl::Span<const uint8_t>(ptr.get(), der.size());
+
+  bssl::UniquePtr<OCSP_BASICRESP> basic_response;
+  bssl::UniquePtr<STACK_OF(X509)> server_cert_chain;
+  ExtractAndVerifyBasicOCSP(malformed_der,
+                            OCSP_RESPONSE_STATUS_SUCCESSFUL,
+                            ca_cert,
+                            server_cert,
+                            0,
+                            &basic_response,
+                            &server_cert_chain);
+}
+
+// Test OCSP response signed by the correct responder certificate, but not for
+// the requested certificate. (So this would be a completely valid response to a
+// different OCSP request for the other certificate.)
+TEST(OCSPTest, TestNotRequestedOCSP) {
+  bssl::UniquePtr<OCSP_BASICRESP> basic_response;
+  bssl::UniquePtr<STACK_OF(X509)> server_cert_chain;
+  ExtractAndVerifyBasicOCSP(ocsp_response_der,
+                            OCSP_RESPONSE_STATUS_SUCCESSFUL,
+                            ca_cert,
+                            server_ecdsa_cert,
+                            1,
+                            &basic_response,
+                            &server_cert_chain);
+
+  X509 *subject = sk_X509_value(server_cert_chain.get(), 0);
+  X509 *issuer = sk_X509_value(server_cert_chain.get(), 1);
+  int status = 0;
+  int reason = 0;
+  // Convert issuer certificate to |OCSP_CERTID|
+  bssl::UniquePtr<OCSP_CERTID> cert_id = bssl::UniquePtr<OCSP_CERTID>(OCSP_cert_to_id(EVP_sha1(), subject, issuer));
+  ASSERT_TRUE(cert_id);
+
+  ASN1_GENERALIZEDTIME *revtime, *thisupd, *nextupd;
+  // Checks revocation status of the response
+  const int ocsp_resp_find_status_res = OCSP_resp_find_status(basic_response.get(), cert_id.get(), &status, &reason, &revtime, &thisupd, &nextupd);
+  ASSERT_EQ(0, ocsp_resp_find_status_res);
+}
+
+// Test OCSP response where the requested certificate was signed by the OCSP
+// responder, but signed by the wrong responder certificate. However, this incorrect
```

Comment:
Is it an error if the certificate is not present in the CERT field of the response? 

I would go with a variant of your explanation instead of "wrong". It better describes what the failure case is.

Reply:
It should be an error if the certificate isn't present since the verification method uses the `certs` in `OCSP_BASICRESP` to verify the signer. We'll try to test that in CryptoAlg-777.

I'll update the comments to make it clearer 👍 .